### ISSUE TITLE
Error when replica installing

### ISSUE THREAD

**Issue by alexandrovas** [freeipa/freeipa-container#374]
After upgrade my FreeIPA environment to image with tag centos-8-4.8.7 new replica installation crashes with this error:
```
Unconfigured automount client failed: CalledProcessError(Command ['/usr/sbin/ipa-client-automount', '--uninstall', '--debug'] returned non-zero exit status 1: 'Loading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nLoading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nLoading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nStarting external process\nargs=[\'/usr/bin/authselect\', \'disable-feature\', \'with-custom-automount\']\nProcess finished, return code=4\nstdout=\nstderr=Running authselect disable-feature with-custom-automount\n[error] [/etc/authselect/nsswitch.conf] is not a regular file!\n[error] Link [/etc/nsswitch.conf] does not point to [/etc/authselect/nsswitch.conf]\n[error] [/etc/nsswitch.conf] was not created by authselect!\n[error] Unexpected changes to the configuration were detected.\n[error] Refusing to activate profile unless those changes are removed or overwrite is requested.\nUnable to disable feature [17]: File exists\n\nTraceback (most recent call last):\n  File "/usr/sbin/ipa-client-automount", line 27, in <module>\n    main()\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 600, in main\n    configure_automount()\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 473, in configure_automount\n    return uninstall(fstore, statestore)\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 322, in uninstall\n    tasks.disable_ldap_automount(statestore)\n  File "/usr/lib/python3.6/site-packages/ipaplatform/redhat/tasks.py", line 759, in disable_ldap_automount\n    ipautil.run(authselect_cmd)\n  File "/usr/lib/python3.6/site-packages/ipapython/ipautil.py", line 598, in run\n    p.returncode, arg_string, output_log, error_log\nipapython.ipautil.CalledProcessError: CalledProcessError(Command [\'/usr/bin/authselect\', \'disable-feature\', \'with-custom-automount\'] returned non-zero exit status 4: \'Running authselect disable-feature with-custom-automount\\n[error] [/etc/authselect/nsswitch.conf] is not a regular file!\\n[error] Link [/etc/nsswitch.conf] does not point to [/etc/authselect/nsswitch.conf]\\n[error] [/etc/nsswitch.conf] was not created by authselect!\\n[error] Unexpected changes to the configuration were detected.\\n[error] Refusing to activate profile unless those changes are removed or overwrite is requested.\\nUnable to disable feature [17]: File exists\\n\')\n')
```

Full docker log:
```
freeipa    | Configuring client side components
freeipa    | stname: ipa02.shared.pcbl.local
freeipa    | Realm: PCBL.LOCAL
freeipa    | DNS Domain: pcbl.local
freeipa    | IPA Server: ipa02.prod-ins2.pcbl.local
freeipa    | BaseDN: dc=pcbl,dc=local
freeipa    | Synchronizing time
freeipa    | No SRV records of NTP servers found and no NTP server or pool address was provided.
freeipa    | Attempting to sync time with chronyc.
freeipa    | Time synchronization was successful.
freeipa    | Successfully retrieved CA cert
freeipa    |     Subject:     CN=Certificate Authority,O=PCBL.LOCAL
freeipa    |     Issuer:      CN=Certificate Authority,O=PCBL.LOCAL
freeipa    |     Valid From:  2020-07-20 16:36:54
freeipa    |     Valid Until: 2040-07-20 16:36:54
freeipa    |
freeipa    | Enrolled in IPA realm PCBL.LOCAL
freeipa    | Created /etc/ipa/default.conf
freeipa    | Configured sudoers in /data/etc/authselect/user-nsswitch.conf
freeipa    | Configured /etc/sssd/sssd.conf
freeipa    | Configured /etc/krb5.conf for IPA realm PCBL.LOCAL
freeipa    | Systemwide CA database updated.
freeipa    | SSSD enabled
freeipa    | Configured /etc/openldap/ldap.conf
freeipa    | Configuring pcbl.local as NIS domain.
freeipa    | Client configuration complete.
freeipa    | The ipa-client-install command was successful
freeipa    | Lookup failed: Preferred host ipa02.shared.pcbl.local does not provide DNS.
freeipa    | Unconfigured automount client failed: CalledProcessError(Command ['/usr/sbin/ipa-client-automount', '--uninstall', '--debug'] returned non-zero exit status 1: 'Loading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nLoading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nLoading StateFile from \'/var/lib/ipa-client/sysrestore/sysrestore.state\'\nStarting external process\nargs=[\'/usr/bin/authselect\', \'disable-feature\', \'with-custom-automount\']\nProcess finished, return code=4\nstdout=\nstderr=Running authselect disable-feature with-custom-automount\n[error] [/etc/authselect/nsswitch.conf] is not a regular file!\n[error] Link [/etc/nsswitch.conf] does not point to [/etc/authselect/nsswitch.conf]\n[error] [/etc/nsswitch.conf] was not created by authselect!\n[error] Unexpected changes to the configuration were detected.\n[error] Refusing to activate profile unless those changes are removed or overwrite is requested.\nUnable to disable feature [17]: File exists\n\nTraceback (most recent call last):\n  File "/usr/sbin/ipa-client-automount", line 27, in <module>\n    main()\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 600, in main\n    configure_automount()\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 473, in configure_automount\n    return uninstall(fstore, statestore)\n  File "/usr/lib/python3.6/site-packages/ipaclient/install/ipa_client_automount.py", line 322, in uninstall\n    tasks.disable_ldap_automount(statestore)\n  File "/usr/lib/python3.6/site-packages/ipaplatform/redhat/tasks.py", line 759, in disable_ldap_automount\n    ipautil.run(authselect_cmd)\n  File "/usr/lib/python3.6/site-packages/ipapython/ipautil.py", line 598, in run\n    p.returncode, arg_string, output_log, error_log\nipapython.ipautil.CalledProcessError: CalledProcessError(Command [\'/usr/bin/authselect\', \'disable-feature\', \'with-custom-automount\'] returned non-zero exit status 4: \'Running authselect disable-feature with-custom-automount\\n[error] [/etc/authselect/nsswitch.conf] is not a regular file!\\n[error] Link [/etc/nsswitch.conf] does not point to [/etc/authselect/nsswitch.conf]\\n[error] [/etc/nsswitch.conf] was not created by authselect!\\n[error] Unexpected changes to the configuration were detected.\\n[error] Refusing to activate profile unless those changes are removed or overwrite is requested.\\nUnable to disable feature [17]: File exists\\n\')\n')
freeipa    | Unenrolling client from IPA server
freeipa    | Removing Kerberos service principals from /etc/krb5.keytab
freeipa    | Disabling client Kerberos and LDAP configurations
freeipa    | Restoring user-nsswitch.conf
freeipa    | Failed to remove krb5/LDAP configuration: CalledProcessError(Command ['/usr/bin/authselect', 'backup-restore', 'pre_ipaclient_20210208101309'] returned non-zero exit status 1: 'Running authselect backup-restore pre_ipaclient_20210208101309\n[error] File [/var/lib/authselect/backups/pre_ipaclient_20210208101309/system-auth] should exist but is missing. It is not safe to delete [/etc/pam.d/system-auth]. Aborting.\n[error] Unable to restore [pre_ipaclient_20210208101309] [1]: Operation not permitted\nUnable to restore backup [pre_ipaclient_20210208101309] [1]: Operation not permitted\n')
freeipa    | The ipa-client-install command failed. See /var/log/ipaclient-uninstall.log for more information
freeipa    | Connection check failed!
freeipa    | See /var/log/ipareplica-conncheck.log for more information.
freeipa    | If the check results are not valid it can be skipped with --skip-conncheck parameter.
freeipa    | The ipa-replica-install command failed. See /var/log/ipareplica-install.log for more information
```

**Comment by adelton** [comment]
What `docker run` options and/or `IPA_SERVER_INSTALL_OPTS` environment variable value did you run the replica container?

**Comment by alexandrovas** [comment]
I use file with this content:
```
--admin-password=<password>
--no-host-dns
--setup-dns
--no-forwarders
--setup-ca
--no-ssh
--no-sshd
--unattended
```

**Comment by adelton** [comment]
To test your setup with `ipa-replica-install-options`, I used
```
--- a/tests/run-master-and-replica.sh
+++ b/tests/run-master-and-replica.sh
@@ -69,7 +69,6 @@ function run_ipa_container() {
        HOSTNAME=ipa.example.test
        if [ "$N" == "freeipa-replica" ] ; then
                HOSTNAME=replica.example.test
-               VOLUME=/tmp/freeipa-test-$$/data-replica
        fi
        mkdir -p $VOLUME
        OPTS=
@@ -79,13 +78,16 @@ function run_ipa_container() {
        if [ -n "$seccomp" ] ; then
                OPTS="$OPTS --security-opt=seccomp:$seccomp"
        fi
+       if [ "$N" == "freeipa-master" ] ; then
+               OPTS="$OPTS -e PASSWORD=Secret123"
+       fi
        (
        set -x
        umask 0
        $docker run $readonly_run -d --name "$N" -h $HOSTNAME \
                $OPTS \
                -v $VOLUME:/data:Z $DOCKER_RUN_OPTS \
-               -e PASSWORD=Secret123 "$IMAGE" "$@"
+               "$IMAGE" "$@"
        )
        wait_for_ipa_container "$N" "$@"
 }
@@ -176,7 +178,19 @@ SETUP_CA=--setup-ca
 if [ $(( $RANDOM % 2 )) == 0 ] ; then
        SETUP_CA=
 fi
-run_ipa_container $IMAGE freeipa-replica no-exit ipa-replica-install -U --principal admin $SETUP_CA --no-ntp
+VOLUME=/tmp/freeipa-replica-1/data
+mkdir -p $VOLUME
+cat > $VOLUME/ipa-replica-install-options <<EOF
+--admin-password=Secret123
+--no-host-dns
+--setup-dns
+--no-forwarders
+--setup-ca
+--no-ssh
+--no-sshd
+--unattended
+EOF
+run_ipa_container $IMAGE freeipa-replica no-exit
 date
 if $docker diff freeipa-master | tee /dev/stderr | grep -Evf tests/docker-diff-ipa.out | grep . ; then
        exit 1
```
and run
```
tests/run-master-and-replica.sh freeipa/freeipa-server:centos-8-4.8.7
```

It passed and the output around that critical section
```
freeipa    | Client configuration complete.
freeipa    | The ipa-client-install command was successful
freeipa    | Lookup failed: Preferred host ipa02.shared.pcbl.local does not provide DNS.
freeipa    | Unconfigured automount client failed: CalledProcessError(Command ['/usr/sbin/ipa-client-automount', '--uninstall', '--debug'] returned non-zero exit status 1 ...
freeipa    | Unenrolling client from IPA server
```
is
```
Client configuration complete.
The ipa-client-install command was successful
Lookup failed: Preferred host replica.example.test does not provide DNS.
  [25/42]: enabling SASL mapping fallback
  [26/42]: restarting directory server
```
So I don't see that `ipa-client-automount` executed or failed.

Can you try this test to make sure it passes in your environment? The format of your output suggests you start the container with something like `docker-compose` and we want to isolate where the difference in behaviour comes from.

**Comment by alexandrovas** [comment]
@adelton 

You are rigth, I used docker-compose:
```yaml
version: "3.3"

services:
  freeipa:
    image: freeipa/freeipa-server:centos-8-4.8.7
    container_name: freeipa
    restart: always
    network_mode: host
    hostname: ipa02.shared.pcbl.local
    sysctls:
      - net.ipv6.conf.all.disable_ipv6=0
    cap_add:
      - SYS_TIME
    volumes:
      - ./data:/data:Z
      - ./conf/ipa-options-ext.conf:/etc/named/ipa-options-ext.conf:ro
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
      - /dev/urandom:/dev/random:ro
    tmpfs:
      - /run
      - /tmp
    environment:
      TZ: Europe/Moscow
      IPA_SERVER_HOSTNAME: ipa02.shared.pcbl.local
      IPA_SERVER_IP: 10.100.0.12
    extra_hosts:
      - "ipa02.shared.pcbl.local:10.100.0.12"
      - "ipa02:10.100.0.12"
```

**Comment by alexandrovas** [comment]
@adelton 
I ran `run-master-and-replica.sh` the test script in my environment. Test passed

**Comment by alexandrovas** [comment]
I have also tried installing the replica with a command like this:
```bash
docker run -e IPA_SERVER_IP=10.100.0.12 -e IPA_SERVER_HOSTNAME=ipa02.shared.pcbl.local -e TZ=Europe/Moscow -v /opt/freeipa/data:/data:Z -v /sys/fs/cgroup:/sys/fs/cgroup:ro -v /dev/urandom:/dev/random:ro -h ipa02.shared.pcbl.local --name freeipa --network host freeipa/freeipa-server:centos-8-4.8.7
```
with a previously placed similar file `/opt/data/ipa-replica-install-options`

The installation error is the same.

**Comment by alexandrovas** [comment]
I found this in `ipareplica-install.log`:
```
...
2021-02-09T14:13:38Z DEBUG Starting external process
2021-02-09T14:13:38Z DEBUG args=['/usr/sbin/ipa-replica-conncheck', '--master', 'ipa01.prod-ins2.pcbl.local', '--auto-master-check', '--realm', 'PCBL.LOCAL', '--hostname', 'ipa02.shared.pcbl.local', '--password', XXXXXXXX, '--ca-cert-file', '/etc/ipa/ca.crt']
2021-02-09T14:14:06Z DEBUG Process finished, return code=1
2021-02-09T14:14:06Z DEBUG stdout=
2021-02-09T14:14:06Z DEBUG stderr=Check connection from replica to remote master 'ipa01.prod-ins2.pcbl.local':
   Directory Service: Unsecure port (389): OK
   Directory Service: Secure port (636): OK
   Kerberos KDC: TCP (88): OK
   Kerberos Kpasswd: TCP (464): OK
   HTTP Server: Unsecure port (80): OK
   HTTP Server: Secure port (443): OK

The following list of ports use UDP protocol and would need to be
checked manually:
   Kerberos KDC: UDP (88): SKIPPED
   Kerberos Kpasswd: UDP (464): SKIPPED

Connection from replica to master is OK.
Start listening on required ports for remote master check
Get credentials to log in to remote master
Check RPC connection to remote master
Execute check on remote master
ERROR: Remote master check failed with following error message(s):
an internal error has occurred

2021-02-09T14:14:06Z DEBUG Starting external process
2021-02-09T14:14:06Z DEBUG args=['/usr/sbin/ipa-client-install', '--unattended', '--uninstall']
...
```

**Comment by rcritten** [comment]
Look in /var/log/ipaclient-install.log for futher details.

**Comment by alexandrovas** [comment]
Part of `ipaclient-install.log` for this time range:
```
2021-02-09T14:13:36Z DEBUG Starting external process
2021-02-09T14:13:36Z DEBUG args=['/bin/systemctl', 'restart', 'nis-domainname.service']
2021-02-09T14:13:36Z DEBUG Process finished, return code=0
2021-02-09T14:13:36Z DEBUG stdout=
2021-02-09T14:13:36Z DEBUG stderr=
2021-02-09T14:13:36Z DEBUG Starting external process
2021-02-09T14:13:36Z DEBUG args=['/bin/systemctl', 'is-active', 'nis-domainname.service']
2021-02-09T14:13:36Z DEBUG Process finished, return code=0
2021-02-09T14:13:36Z DEBUG stdout=active

2021-02-09T14:13:36Z DEBUG stderr=
2021-02-09T14:13:36Z DEBUG Restart of nis-domainname.service complete
2021-02-09T14:13:36Z INFO Client configuration complete.
2021-02-09T14:13:36Z INFO The ipa-client-install command was successful
2021-02-09T14:14:07Z DEBUG Loading StateFile from '/var/lib/ipa-client/sysrestore/sysrestore.state'
2021-02-09T14:14:07Z DEBUG Loading StateFile from '/var/lib/ipa-client/sysrestore/sysrestore.state'
2021-02-09T14:14:07Z DEBUG Loading StateFile from '/var/lib/ipa-client/sysrestore/sysrestore.state'
2021-02-09T14:14:07Z DEBUG Starting external process
2021-02-09T14:14:07Z DEBUG args=['/usr/bin/authselect', 'disable-feature', 'with-custom-automount']
2021-02-09T14:14:07Z DEBUG Process finished, return code=4
```

**Comment by rcritten** [comment]
That's not enough context. The return code means the uninstall failed but this blew up during the install as well so something really went wrong. What you're looking for is not at the end of the log.

**Comment by alexandrovas** [comment]
This problem was related to network AСLs in my environment. Sorry... Issue can be closed.

**Comment by adelton** [comment]
Thanks for the followup.

**Comment by adelton** [comment]
For the record, the `ipa-client-automount --uninstall --debug` gets called when something in `ipa-replica-install` fails because the installer tries to revert the changes it thinks it did. We've seen the same on Fedora 32 and 33 when master did not set resolv.conf correctly and then could resolve the IPA-enrolled replica and connection check failed. Those Fedora issues got now fixed in master in 632d3b3028526af2c3765f3a838c200ddd46a1cc.
